### DISCUSSION TITLE
Apple Intel macOS Ventura 13.0.1: unable to install ruby 2.6.9

### DISCUSSION THREAD

**Discussion by Zandelok** [rbenv/ruby-build#2090] [Build failures] [answered]
### Steps to reproduce the behaviour
1. `rbenv install 2.6.9`
2. `RUBY_CFLAGS="-Wno-error=implicit-function-declaration" rbenv install 2.6.9`
3. `RUBY_CFLAGS=-DUSE_FFI_CLOSURE_ALLOC arch -x86_64 rbenv install 2.6.9`

Every time I install 2.6.9 I get `BUILD FAILED (macOS 13.0.1 using ruby-build 20221101)`

FYI: 3.1.0 installed successfully

### Expected
`Installed ruby-2.6.9 to ~/.rbenv/versions/2.6.9`

### Actual behavior
<img width="1537" alt="image" src="https://user-images.githubusercontent.com/70264606/201873072-d134232b-fa85-49ae-8429-357257259080.png">

### Logs

Full log file: https://gist.github.com/Zandelok/844e89b7c964a751c504867745b694ef

**Comment by mislav** [comment]
Not sure why this failed, but please note that the version you're trying to compile is ancient (in relative terms) and your operating system is very new. You might encounter incompatibilities and should upgrade Ruby in your app to at least v3.0.

You could try building 2.6.10 with `RUBY_CONFIGURE_OPTS="--disable-shared"`, which will skip building the shared library which is for most intents and purposes optional. If none of the ruby gems in your app need the shared library, you might get away with that. It looks like your build process trips while building the shared library.

**Reply by SampsonCrowley** [reply]
@mislav people who are working on older ruby versions aren't always doing it by choice. I cant just stop working until the company I work for decides they're ok with upgrading (which I've tried to convince them of multiple times)

Legacy code is real, and sometimes unavoidable, and "you should upgrade" statements aren't always exactly helpful. ruby 2.6 (which is old, yes, but not ancient; 2.4 is ancient) compiled just fine before ventura, and then it just completely shits the bed with xcode 14. all people like us are looking for is a way to get back to work after a system updated itself

**Reply by SampsonCrowley** [reply]
and thanks for the pointer toward disabling shared libraries. didn't know that was an option

**Reply by mislav** [reply]
@SampsonCrowley I understand that it's not always easy or immediately available to update a Ruby version in your project. My statement of the version being “ancient” should have been more precise: Ruby 2.6 is now EOL (end of life) and no longer receiving updates. What that means is that if newer macOS (or any other system) introduces changes that somehow break Ruby compilation, Ruby 2.6 will never receive any fixes to make that better. The only path forward for you will be either 1) applying unofficial patches to Ruby 2.6 or; 2) upgrading Ruby.

**Reply by pykenny** [reply]
`--disable-shared` flag + suppressing compile warnings (that appeared in error log) doesn't work, though (Ventura + XCode CLI Tool 14.2 + Intel):

```shell
CFLAGS="-Wno-compound-token-split-by-macro -Wno-pointer-to-enum-cast -Wno-string-plus-int" RUBY_CONFIGURE_OPTS="--enable-shared" rbenv install 2.6.10
```
➡️ 
```
linking encoding encdb.bundle
Undefined symbols for architecture x86_64:
  "_rb_enc_set_base", referenced from:
      _Init_encdb in encdb.o
  "_rb_enc_set_dummy", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_alias", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_declare", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_dummy", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_replicate", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_set_unicode", referenced from:
      _Init_encdb in encdb.o
ld: symbol(s) not found for architecture x86_64

...

linking transcoder transdb.bundle
Undefined symbols for architecture x86_64:
  "_rb_declare_transcoder", referenced from:
      _Init_transdb in transdb.o
ld: symbol(s) not found for architecture x86_64
```

**Reply by pykenny** [reply]
(Finally ended up with using [Homebrew's prebuilt binaries](https://formulae.brew.sh/formula/ruby@2.6) after several tries.)

**Comment by kvoider-mdsol** [comment]
I have the same issue and found this [thread](https://bugs.ruby-lang.org/issues/18912?tab=history). But nothing helps me.

**Reply by Zandelok** [reply]
I finally solved this problem. Reinstalling the Xcode helped me (remove it and install last version from App Store)

**Reply by kvoider-mdsol** [reply]
not for me((
I did it twice.

Do you have M1 or Intel chip?

**Reply by Zandelok** [reply]
Intel (i mentioned it on title)

**Reply by kvoider-mdsol** [reply]
Could you write what commands have you fired for installing ruby please?

**Reply by Zandelok** [reply]
`rbenv install 2.6.9`
Nothing special, all the problems were solved after reinstallation (just for me).
My colleague has the same problem with rvm, reinstalling xcode also helped him.

**Reply by damdl96** [reply]
I was having the same issue on an M2 Macbook Pro through a Rosetta terminal, this is what I did to solve it:
```
CFLAGS='-Wno-compound-token-split-by-macro -Wno-pointer-to-enum-cast -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix' rbenv install 2.6.9
```
Make sure to execute it all in the same command otherwise the CFLAGS won't apply.

**Comment by Zandelok** [comment, marked as answer]
Reinstalling the Xcode helped me (remove it and install last version from App Store)

**Reply by gabrieelgomez** [reply]
Yes, this worked for me too!

**Reply by AndreyBartnovskiy** [reply]
Yes, thx this worked. What are you need to do:
1. destroy Xcode CLT with command: `sudo rm -rf /Library/Developer/CommandLineTools`
2. reinstall Xcode: `xcode-select --install`
3. try to install rbenv with flag like this: `RUBY_CFLAGS="-Wno-error=implicit-function-declaration" rbenv install 2.5.1`
Hope this helps you :D

**Comment by mikes-checkr** [comment]
EDIT: The answer given is correct for Intel Macs, which is the scope of this Issue/Thread. This fix will not work for M1 Macs whether running natively or under Rosetta. Leaving original post below unedited for context of future readers.

The answer given is incorrect. Ruby <2.7.7 cannot be built on OSX Ventura due to incompatibilities with the xcode libraries in xcode 14.0. The initial bugs are reported here https://bugs.ruby-lang.org/issues/18912 and here https://bugs.ruby-lang.org/issues/19005

This is also a problem in the latest version of Monterey (12.6), as it will automatically install xcode 14. In OSX Monterey, it is possible to downgrade to an earlier version of the xcode command line tools, but in Ventura you're stuck with 14.0+.

This issue affected many versions of Ruby when xcode 14 was released, for instance ruby 2.7.6 and earlier versions of 2.7 will also not build on Ventura. This issue was fixed and backported to all supported Ruby versions: For instance here is the backport for 2.7 that was released in 2.7.7: https://github.com/ruby/ruby/pull/6442/commits/0525a1ecd92cf72dab817b8852c809deb4abab0f

Since ruby 2.6 is EOL, the fix was not backported to 2.6 and it remains, to the best of my knowledge, not possible to build it (without changes to ruby) on Ventura. If you want to throw money at the problem, apparently ruby-on-mac (rom) has a fix for it (likely either by forking ruby 2.6.10 and backporting the 2.7.7 fix themselves, or by distributing a pre-built binary). See article here: https://www.rubyonmac.dev/how-to-install-ruby-2-6-on-macos-13-ventura

To be clear the issue is with BUILDING ruby not RUNNING ruby. If you `rbenv install 2.6.10` on Monterey 12.5 with Xcode 13.4.0 then upgrade to Ventura 13.1 with Xcode 14.0, your existing ruby distribution should continue to work fine. But if you ever try to rbenv install another version of ruby <2.7.7 once you've upgraded to ventura, that installation will also fail.

**Reply by Zandelok** [reply]
As I mentioned in 'answer', it helped me. Not read it somewhere, not heard. This solution solved my issue with ruby 2.6.9. I didn't have any ruby version on my Mac. I also installed 2.7.0 version without problems

**Reply by Zandelok** [reply]
And this solution also helped @gabrieelgomez (check comments below answer)

**Reply by mikes-checkr** [reply]
Hrm... Its definitley not working here... maybe it only works for certain architectures/compiler flags? I see your configured for an X86 architecture. Are you on a M1 w/Rosetta or an Intel chip?

**Reply by mikes-checkr** [reply]
Also I see you did an App Store install as opposed to an xcode-select install - something else for me to try. What version of Xcode did you run on when things ultimately worked?

**Reply by hsbt** [reply]
@mikes-checkr Your understanding is correct. But I didn't investigate Intel CPU, Ventura and Xcode 14 environment. we might not have any issues to build old versions.

**Reply by Zandelok** [reply]
@mikes-checkr Intel (mentioned in Title)

**Reply by SampsonCrowley** [reply]
You're both wrong and right. The permanency of the problem only applies to M1 users.

The problem, which the title is for, isn't specific to Intel users, and shouldnt be relegated to it. The fix, however is. M1 and Intel versions have different binaries and libraries. On Intel it fixes the problem because things just need to be relinked after upgrade.

On M1, building is permanently broken, at least until apple stops treating their users like legacy code doesn't exist and backwards compatibility isn't important. Which they won't ever do, because in their eyes, everything they do is right, no matter how much pain it causes for the people that actually use the software.

**Comment by mikes-checkr** [comment]
Thanks @SampsonCrowley, this is what I found over the weekend too. We're on M1's and we were unable to resolve this whether running under Rosetta or natively. Last night, I upgraded my personal mac which is an Intel to Ventura, and reinstalled XCode and that resolved the issue just as @Zandelok suggests. Apologies to @Zandelok for misunderstanding the scope of the fix and suggesting it didn't work. It works just fine - for Intel macs, which is the scope of this thread.

**Reply by prateeksen** [reply]
Passing RUBY_CFLAGS="-w" worked for me on MacOS Ventura M1 Max. 
example: 
`RUBY_CFLAGS="-w" rbenv install 2.4.0`

**Reply by taoxuau** [reply]
> Passing RUBY_CFLAGS="-w" worked for me on MacOS Ventura M2. example: `RUBY_CFLAGS="-w" rbenv install 2.4.0`

Thanks @prateek-ap , this works for me on MacOS Ventura M2. `RUBY_CFLAGS="-w" rbenv install 2.6.9`

**Reply by Kawixh** [reply]
Ok this works for me too I am on Ventura with intel based Macbook 2020 <3

**Reply by odarriba** [reply]
This have just worked for me on M2 Pro, Ventura and Ruby 2.6.6

**Reply by rafaeljustinox** [reply]
Worked on M1 Ruby 2.6.6

**Reply by laterty** [reply]
Ventura 13.3, Intel 2019, ruby 2.6.5, thanks a lot, `RUBY_CFLAGS="-w" asdf install ruby 2.6.5` now working, thanks again
